### PR TITLE
Add Annotations Support for Cilium

### PR DESCRIPTION
**Pull request by mateoflorido**
## Overview
Add annotation support for Cilium network features

## Rationale
To enable the Sunbeam team to configure specific functionalities in Cilium, we are adding one of the requested features (VLANBPFBypass) to the annotations support. This configuration can then be passed during the bootstrap  via the configuration file.

## Related Issues
https://github.com/canonical/k8s-operator/issues/112

### CODE REVIEW

File: src/k8s/pkg/k8sd/features/cilium/internal.go [hunk 1]
```diff
@@ -1,18 +1,78 @@
...
+}
+
+func validateVLANBPFBypass(vlanList string) ([]int, error) {
+	vlanList = strings.TrimSpace(vlanList)
+	// Maintain compatibility with the Cilium chart definition
+	vlanList = strings.Trim(vlanList, "{}")
+	vlans := strings.Split(vlanList, ",")
+
+	// Special case: wildcard "0" allows all VLANs
+	if len(vlans) == 1 && strings.TrimSpace(vlans[0]) == "0" {
```
Comment (bschimke95) [inline src/k8s/pkg/k8sd/features/cilium/internal.go:35]: What happens if a use sets something like this:
```
10,0,11
```
Will cilium handle this ID or will it fail?

File: src/k8s/pkg/k8sd/features/cilium/internal.go [hunk 2]
```diff
@@ -1,18 +1,78 @@
...
+		return []int{0}, nil
+	}
+
+	if len(vlans) > maxVLANTags {
```
Comment (bschimke95) [inline src/k8s/pkg/k8sd/features/cilium/internal.go:39]: Why do we have this limitation? Is this imposed by cilium?

File: src/k8s/pkg/k8sd/features/cilium/internal.go [hunk 3]
```diff
@@ -1,18 +1,78 @@
...
+		vlanID, err := strconv.Atoi(strings.TrimSpace(vlan))
+		if err != nil {
+			return []int{}, fmt.Errorf("failed to parse VLAN tag: %w", err)
+		}
+		if vlanID < minVLANIDValue || vlanID > maxVLANIDValue {
+			return []int{}, fmt.Errorf("VLAN tag must be between 0 and %d", maxVLANIDValue)
+		}
+
+		if _, ok := seenTags[vlanID]; ok {
+			return []int{}, fmt.Errorf("VLAN tag %d is duplicated", vlanID)
```
Comment (bschimke95) [inline src/k8s/pkg/k8sd/features/cilium/internal.go:56]: Up for discussion: Is this really an issue that we should fail on?

**Review by bschimke95 — COMMENTED** [review]
Great work, couple of questions but generally LGTM

File: src/k8s/pkg/k8sd/features/cilium/internal.go (hunk 2 above)
Comment (mateoflorido) [reply src/k8s/pkg/k8sd/features/cilium/internal.go:39]: Yes. Unfortunately, this a limitation imposed by the `cilium-agent`: https://docs.cilium.io/en/stable/configuration/vlan-802.1q/

File: src/k8s/pkg/k8sd/features/cilium/internal.go (hunk 3 above)
Comment (mateoflorido) [reply src/k8s/pkg/k8sd/features/cilium/internal.go:56]: I believe we can relax this check and not fail when we are encountering a duplicate. We could simply avoid including the same value multiple times in the chart configuration value. Wdyt?

File: src/k8s/pkg/k8sd/features/cilium/internal.go (hunk 1 above)
Comment (mateoflorido) [reply src/k8s/pkg/k8sd/features/cilium/internal.go:35]: Cilium will handle that and allow all the VLAN IDs on the interface: https://github.com/cilium/cilium/blob/main/pkg/datapath/linux/config/config.go#L851-L861
@bschimke95 Should I refactor the code to allow the user to provide the wildcard along with other VLAN IDs?

File: src/k8s/pkg/k8sd/features/cilium/internal.go (hunk 3 above)
Comment (bschimke95) [reply src/k8s/pkg/k8sd/features/cilium/internal.go]: sounds good

**Review by bschimke95 — APPROVED** [review]
LGTM

**Pull request merged**